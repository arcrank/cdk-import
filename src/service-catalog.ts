import * as AWS from 'aws-sdk';
import { createAwsClient } from './aws';

export interface IServiceCatalogClient {
  searchProducts(input: AWS.ServiceCatalog.ListPortfoliosInput): Promise<AWS.ServiceCatalog.ListPortfoliosOutput>;
  listProvisioningArtifacts(input: AWS.ServiceCatalog.ListProvisioningArtifactsInput): Promise<AWS.ServiceCatalog.ListProvisioningArtifactsOutput>;
  listLaunchPaths(input: AWS.ServiceCatalog.ListLaunchPathsInput): Promise<AWS.ServiceCatalog.ListLaunchPathsOutput> ;
  describeProvisioningParameters(input: AWS.ServiceCatalog.DescribeProvisioningParametersInput):
  Promise<AWS.ServiceCatalog.DescribeProvisioningParametersOutput>;
}

export class ServiceCatalogClient implements IServiceCatalogClient {
  private readonly sc: AWS.ServiceCatalog;

  constructor() {
    this.sc = createAwsClient(AWS.ServiceCatalog);
  }

  public async searchProducts(input: AWS.ServiceCatalog.SearchProductsInput): Promise<AWS.ServiceCatalog.SearchProductsOutput> {
    return this.sc.searchProducts(input).promise();
  }

  public async listProvisioningArtifacts(input: AWS.ServiceCatalog.ListProvisioningArtifactsInput):
  Promise<AWS.ServiceCatalog.ListProvisioningArtifactsOutput> {
    return this.sc.listProvisioningArtifacts(input).promise();
  }

  public async listLaunchPaths(input: AWS.ServiceCatalog.ListLaunchPathsInput): Promise<AWS.ServiceCatalog.ListLaunchPathsOutput> {
    return this.sc.listLaunchPaths(input).promise();
  }

  public async describeProvisioningParameters(input: AWS.ServiceCatalog.DescribeProvisioningParametersInput):
  Promise<AWS.ServiceCatalog.DescribeProvisioningParametersOutput> {
    return this.sc.describeProvisioningParameters(input).promise();
  }
}

/**
 * Provide query values for a specific provisionable product.
 * If no specific IDs are declared for a product, we will use the set DEFAULT value for the provisioning artifact and launch path.
 */
export interface DescribeProvisionedProductOptions {
  /**
   * A client that performs calls to AWS Service Catalog. You can provide a mock here
   * for testing.
   */
  readonly client?: IServiceCatalogClient;

  /**
   * The product Id..
   */
  readonly productId?: string;

  /**
   * The provisioning artifact Id.
   */
  readonly provisioningArtifactId?: string;

  /**
   * The launch path Id.
   */
  readonly launchPathId?: string;
}

/**
 * Fetches the all the available or queried service catalog product(s) for the caller.
 *
 * @returns list of product view summaries
 */
export async function fetchAvailableProducts(options: DescribeProvisionedProductOptions = {}): Promise<AWS.ServiceCatalog.ProductViewSummaries> {
  const sc = options.client ?? new ServiceCatalogClient();

  const products = [];
  let token;

  do {
    const res: AWS.ServiceCatalog.SearchProductsOutput = await sc.searchProducts({
      PageToken: token,
      ...(options.productId ? { SourceProductId: options.productId } : {}),
    });

    if (res.ProductViewSummaries) {
      products.push(...res.ProductViewSummaries);
    }
    token = res.NextPageToken;
  } while (token);

  return products;
}

/**
 * Fetches the all the available or queried provisioning artifact(s) for the caller.
 *
 * @param productId the product to fetch provisioning artifacts for
 * @returns list of provisioning artifact details for a product
 */
export async function listProvisioningArtifacts(
  productId: string,
  options: DescribeProvisionedProductOptions = {}): Promise<AWS.ServiceCatalog.ProvisioningArtifactDetail[]> {
  const sc = options.client ?? new ServiceCatalogClient();

  const provisioningArtifacts: AWS.ServiceCatalog.ListProvisioningArtifactsOutput = await sc.listProvisioningArtifacts({
    ProductId: productId,
  });

  if ( options.provisioningArtifactId != null ) {
    return provisioningArtifacts.ProvisioningArtifactDetails!.filter(pa => pa.Id == options.provisioningArtifactId);
  } else {
    return provisioningArtifacts.ProvisioningArtifactDetails!;
  }
}

/**
 * Retrieves all the availalbe or queried launch path(s) for a product.
 *
 * @param productId the product to list launch paths for
 * @returns list of launchPathSummaries for a product
 */
export async function listLaunchPaths(
  productId: string,
  options: DescribeProvisionedProductOptions = {}):
  Promise<AWS.ServiceCatalog.LaunchPathSummaries> {
  const sc = options.client ?? new ServiceCatalogClient();

  const launchPaths = [];
  let token;

  do {
    const res: AWS.ServiceCatalog.ListLaunchPathsOutput = await sc.listLaunchPaths({
      ProductId: productId,
      PageToken: token,
    });

    if (res.LaunchPathSummaries) {
      launchPaths.push(...res.LaunchPathSummaries);
    }
    token = res.NextPageToken;
  } while (token);

  if ( options.launchPathId != null ) {
    return launchPaths.filter(lp => lp.Id == options.launchPathId);
  } else {
    return launchPaths;
  }
}

/**
 * Gets the information required to provision an artifact.
 * Includes information on inputs and outputs of the product.
 *
 * @param productId the product to get provisioning parameters for
 * @param provisioningArtifactId the provisioning artifact to get provisioning parameters for
 * @param launchPathId the launch path to get provisioning parameters for
 * @returns the provisoning parameters for an artifact
 */
export async function describeProvisioningParameters(productId: string, provisioningArtifactId: string, launchPathId: string,
  options: DescribeProvisionedProductOptions = {}):
  Promise<AWS.ServiceCatalog.DescribeProvisioningParametersOutput> {

  const sc = options.client ?? new ServiceCatalogClient();

  const provisioningParameters: AWS.ServiceCatalog.DescribeProvisioningParametersOutput = await sc.describeProvisioningParameters({
    ProductId: productId,
    ProvisioningArtifactId: provisioningArtifactId,
    PathId: launchPathId,
  });

  return provisioningParameters;
}