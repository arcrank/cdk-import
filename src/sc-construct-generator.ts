import { pascal, camel } from 'case';
import * as j2j from 'json2jsii';
import { ProductDataAggregate } from './service-catalog';

/**
 * Generator to emit classes and information on provisioning details for a product.
 */
export class ServiceCatalogProvisioningConstructGenerator {
  private readonly productDataAggregate: ProductDataAggregate;
  private readonly hasParameters: boolean;
  private readonly sanitizedTypeName: string;
  private readonly resourceAttributes: string[];
  private readonly constructClassName: string;
  private readonly propsStructName: string;
  private readonly outputKeys: string[];
  public readonly name: string;

  /**
   * @param productData the aggregate details on a product with information on inputs and outputs of the template.
   */
  constructor(productData: ProductDataAggregate) {
    this.productDataAggregate = productData;
    this.sanitizedTypeName = pascal(this.productDataAggregate.product.Name!);
    this.constructClassName = j2j.TypeGenerator.normalizeTypeName(this.sanitizedTypeName);
    this.propsStructName = j2j.TypeGenerator.normalizeTypeName(`${this.constructClassName}Props`);
    this.name = this.sanitizedTypeName;
    this.resourceAttributes = new Array<string>();
    this.outputKeys = this.productDataAggregate.params.ProvisioningArtifactOutputKeys!.map(key => key.Key!);
    this.hasParameters = this.productDataAggregate.params.ProvisioningArtifactParameters!.length > 0;

    const attributeNames: string[] = [
      'provisionedProductId',
      'cloudFormationStackArn',
      'recordId',
      ...this.outputKeys.map(camel),
    ];

    for (const attr of attributeNames) {
      this.resourceAttributes.push(attr);
    }
  }

  /**
   * Render the type into a TypeScript class file defining the props and the wrapped L1 construct AWS::ServiceCatalog::CloudFormationProvisionedProduct.
   *
   * @returns the rendered class file content
   */
  public render(): string {
    const code = new j2j.Code();
    this.emitImports(code);
    this.emitInterface(code);
    this.emitBase(code);
    if (this.hasParameters) {
      this.emitParameterMap(code);
      this.emitDefinitionTypes(code);
    }
    this.emitConstructClass(code);
    return code.render();
  }

  private emitImports(code: j2j.Code) {
    code.line('// Generated by cdk-import');
    code.line("import * as cdk from 'aws-cdk-lib';");
    code.line("import * as constructs from 'constructs';");
    code.line("import * as sc from 'aws-cdk-lib/aws-servicecatalog';");
    code.line();
  }

  private emitInterface(code: j2j.Code) {
    code.openBlock(`export interface I${this.constructClassName} extends cdk.IResource`);
    this.resourceAttributes.forEach(attr => {
      code.line(`readonly ${attr}: string;`);
    });
    code.closeBlock();
    code.line();
  }

  private emitBase(code: j2j.Code) {
    code.openBlock(`abstract class ${this.constructClassName}Base extends cdk.Resource implements I${this.constructClassName}`);
    this.resourceAttributes.forEach(attr => {
      code.line(`public abstract readonly ${attr}: string;`);
    });
    code.closeBlock();
    code.line();
  }

  private emitParameters(code: j2j.Code) {
    code.line('  provisioningParameters:');
    code.line('    Object.entries(props).map(([k, v]) => {');
    code.line('      return {');
    code.line('        key: ParameterKeys.valueOf(k) as string,');
    code.line('        value: v as string,');
    code.line('      }');
    code.line('    }),');
  }

  private emitParameterMap(code: j2j.Code) {
    code.openBlock('enum ParameterKeys');
    this.productDataAggregate.params.ProvisioningArtifactParameters?.forEach(p => {
      code.line(`${camel(p.ParameterKey!)} = '${p.ParameterKey}',`);
    });
    code.closeBlock();
    code.line();
    code.openBlock('namespace ParameterKeys');
    code.line('const reverseMap = new Map<string, ParameterKeys>();');
    code.line('Object.keys(ParameterKeys).forEach((s: string) => {');
    code.line('  const e = (<any>ParameterKeys)[s];');
    code.line('  reverseMap.set(s, e);');
    code.line('});');
    code.line('export function valueOf(str: string) {');
    code.line('  return reverseMap.get(str);');
    code.line('}');
    code.closeBlock();
  }

  private emitDefinitionTypes(code: j2j.Code) {
    code.openBlock(`export interface ${this.constructClassName}Props`);
    this.productDataAggregate.params.ProvisioningArtifactParameters!.forEach(pa => {
      let optional = '';
      code.line('/**');
      (pa.Description != undefined) ? code.line(` * ${pa.Description}`): null;
      if (pa.DefaultValue) {
        optional += '?';
        (pa.ParameterConstraints?.ConstraintDescription != undefined) ? code.line(` * ${pa.ParameterConstraints?.ConstraintDescription}`): null;
        code.line(' *');
        code.line(` * @default ${pa.DefaultValue}`);
      }

      code.line(' */');
      code.line(`readonly ${camel(pa.ParameterKey!)}${optional}: ${pa.ParameterConstraints?.AllowedValues?.length ? pa.ParameterKey : 'string'};`);
      code.line();
    });
    code.closeBlock();
    code.line();
  }

  private emitConstructClass(code: j2j.Code) {
    code.line('/**');
    code.line(` * A Service Catalog CloudFormation \`${this.sanitizedTypeName}\` product.`);
    code.line(' *');
    code.line(' * @cloudformationResource AWS::ServiceCatalog::CloudFormationProvisionedProduct');
    code.line(' * @stability external');
    code.line(' */');
    code.openBlock(`export class ${this.constructClassName} extends ${this.constructClassName}Base`);
    for (const prop of this.resourceAttributes) {
      code.line('/**');
      code.line(` * Attribute \`${this.sanitizedTypeName}.${prop}\``);
      code.line(' */');
      code.line(`public readonly ${camel(prop)}: string;`);
    }
    code.line('private readonly provisionedProduct: sc.CfnCloudFormationProvisionedProduct;');
    code.line();

    code.line('/**');
    code.line(` * Create a new Service Catalog \`${this.sanitizedTypeName}\` product.`);
    code.line(' *');
    code.line(' * @param scope - scope in which this resource is defined');
    code.line(' * @param id    - scoped id of the resource');
    if (this.hasParameters) {
      code.line(' * @param props - resource properties');
    }
    code.line(' */');
    if (this.hasParameters) {
      code.openBlock(`constructor(scope: constructs.Construct, id: string, props: ${this.propsStructName})`);
    } else {
      code.openBlock('constructor(scope: constructs.Construct, id: string)');
    }
    code.line('super(scope, id);');
    code.line('');
    code.line("this.provisionedProduct = new sc.CfnCloudFormationProvisionedProduct(this, 'Resource', {");
    code.line('  provisionedProductName: this.node.id,');
    code.line(`  provisioningArtifactName: '${this.productDataAggregate.provisioningArtifact.Name!}',`);
    code.line(`  productName: '${this.productDataAggregate.product.Name!}',`);

    // generate provisioning parameters from props to avoid redundant defaults
    if (this.hasParameters) {
      this.emitParameters(code);
    };
    code.line('});');
    code.line();
    code.line('this.provisionedProductId = this.provisionedProduct.ref;');
    code.line("this.cloudFormationStackArn = cdk.Token.asString(cdk.Fn.getAtt(this.provisionedProduct.logicalId, 'CloudformationStackArn'));");
    code.line("this.recordId = cdk.Token.asString(cdk.Fn.getAtt(this.provisionedProduct.logicalId, 'RecordId'));");
    this.outputKeys?.forEach(output => {
      code.line(`this.${camel(output)} = cdk.Token.asString(cdk.Fn.getAtt(this.provisionedProduct.logicalId, 'Outputs.${output}'));`);
    });
    code.closeBlock();

    // Close construct class
    code.closeBlock();
  }
}