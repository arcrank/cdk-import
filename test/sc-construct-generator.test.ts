import { spawnSync } from 'child_process';
import * as fs from 'fs';
import { basename, join } from 'path';
import * as testee from '../src/sc-construct-generator';
import { ProductDataAggregate } from '../src/service-catalog';

const SAMPLES_DIR = join(__dirname, 'fixtures', 'sc-samples');
const SAMPLE_FILES = fs.readdirSync(SAMPLES_DIR);
const JSII = require.resolve('jsii/bin/jsii');
const PKG = JSON.parse(fs.readFileSync(join(__dirname, '..', 'package.json'), 'utf-8'));
const CORE_VERSION = PKG.devDependencies['aws-cdk-lib'];
const CONSTRUCTS_VERSION = PKG.devDependencies.constructs;

test.each(SAMPLE_FILES)('generate %s', async fixture => {
  const productDataAggregate = JSON.parse(fs.readFileSync(join(SAMPLES_DIR, fixture), { encoding: 'utf8' })) as ProductDataAggregate;
  const gen = new testee.ServiceCatalogProvisioningConstructGenerator(productDataAggregate);
  const sourceFile = gen.render();

  //console.log(sourceFile);
  expect(sourceFile).toMatchSnapshot();

  // now compile the generated code to make sure its valid
  const workdir = join(process.cwd(), '.tmp', basename(fixture, '.json'));
  fs.mkdirSync(workdir, { recursive: true });
  fs.writeFileSync(join(workdir, 'index.ts'), sourceFile);
  fs.writeFileSync(join(workdir, 'package.json'), JSON.stringify({
    name: 'test',
    main: 'index.js',
    types: 'index.d.ts',
    jsii: {
      outdir: 'dist',
    },
    version: '1.0.0',
    author: {
      name: 'AWS Service Catalog',
      email: 'b@rubble.com',
      url: 'http://barnyrubble.tumblr.com/',
    },
    license: 'UNLICENSED',
    peerDependencies: {
      'aws-cdk-lib': CORE_VERSION,
      'constructs': CONSTRUCTS_VERSION,
    },
    repository: {
      type: 'git',
      url: 'https://github.com/my/repo.git',
    },
  }, undefined, 2));

  console.log(`Compiling ${workdir}`);
  const ret = spawnSync(JSII, { cwd: workdir, stdio: 'inherit' });
  if (ret.error || ret.status !== 0) {
    throw new Error('compilation failed');
  }

});
