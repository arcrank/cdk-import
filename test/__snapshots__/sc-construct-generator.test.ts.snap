// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate ai-health.json 1`] = `
"// Generated by cdk-import
import * as cdk from 'aws-cdk-lib';
import * as constructs from 'constructs';
import * as sc from 'aws-cdk-lib/aws-servicecatalog';

export interface IAiPoweredHealthDataMasking extends cdk.IResource {
  readonly provisionedProductId: string;
  readonly cloudFormationStackArn: string;
  readonly recordId: string;
  readonly apiGatewayInvokeURL: string;
  readonly apiGatewayId: string;
  readonly apiGatewayStage: string;
  readonly imageBucketName: string;
  readonly logBucketName: string;
  readonly imageResourceId: string;
  readonly imageTextResourceId: string;
  readonly imagePhiResourceId: string;
  readonly imageMaskResourceId: string;
  readonly textPhiResourceId: string;
  readonly textMaskResourceId: string;
}

abstract class AiPoweredHealthDataMaskingBase extends cdk.Resource implements IAiPoweredHealthDataMasking {
  public abstract readonly provisionedProductId: string;
  public abstract readonly cloudFormationStackArn: string;
  public abstract readonly recordId: string;
  public abstract readonly apiGatewayInvokeURL: string;
  public abstract readonly apiGatewayId: string;
  public abstract readonly apiGatewayStage: string;
  public abstract readonly imageBucketName: string;
  public abstract readonly logBucketName: string;
  public abstract readonly imageResourceId: string;
  public abstract readonly imageTextResourceId: string;
  public abstract readonly imagePhiResourceId: string;
  public abstract readonly imageMaskResourceId: string;
  public abstract readonly textPhiResourceId: string;
  public abstract readonly textMaskResourceId: string;
}

/**
 * A Service Catalog CloudFormation \`AIPoweredHealthDataMasking\` product.
 *
 * @cloudformationResource AWS::ServiceCatalog::CloudFormationProvisionedProduct
 * @stability external
 */
export class AiPoweredHealthDataMasking extends AiPoweredHealthDataMaskingBase {
  /**
   * Attribute \`AIPoweredHealthDataMasking.provisionedProductId\`
   */
  public readonly provisionedProductId: string;
  /**
   * Attribute \`AIPoweredHealthDataMasking.cloudFormationStackArn\`
   */
  public readonly cloudFormationStackArn: string;
  /**
   * Attribute \`AIPoweredHealthDataMasking.recordId\`
   */
  public readonly recordId: string;
  /**
   * Attribute \`AIPoweredHealthDataMasking.apiGatewayInvokeURL\`
   */
  public readonly apiGatewayInvokeURL: string;
  /**
   * Attribute \`AIPoweredHealthDataMasking.apiGatewayId\`
   */
  public readonly apiGatewayId: string;
  /**
   * Attribute \`AIPoweredHealthDataMasking.apiGatewayStage\`
   */
  public readonly apiGatewayStage: string;
  /**
   * Attribute \`AIPoweredHealthDataMasking.imageBucketName\`
   */
  public readonly imageBucketName: string;
  /**
   * Attribute \`AIPoweredHealthDataMasking.logBucketName\`
   */
  public readonly logBucketName: string;
  /**
   * Attribute \`AIPoweredHealthDataMasking.imageResourceId\`
   */
  public readonly imageResourceId: string;
  /**
   * Attribute \`AIPoweredHealthDataMasking.imageTextResourceId\`
   */
  public readonly imageTextResourceId: string;
  /**
   * Attribute \`AIPoweredHealthDataMasking.imagePhiResourceId\`
   */
  public readonly imagePhiResourceId: string;
  /**
   * Attribute \`AIPoweredHealthDataMasking.imageMaskResourceId\`
   */
  public readonly imageMaskResourceId: string;
  /**
   * Attribute \`AIPoweredHealthDataMasking.textPhiResourceId\`
   */
  public readonly textPhiResourceId: string;
  /**
   * Attribute \`AIPoweredHealthDataMasking.textMaskResourceId\`
   */
  public readonly textMaskResourceId: string;
  private readonly provisionedProduct: sc.CfnCloudFormationProvisionedProduct;

  /**
   * Create a new Service Catalog \`AIPoweredHealthDataMasking\` product.
   *
   * @param scope - scope in which this resource is defined
   * @param id    - scoped id of the resource
   */
  constructor(scope: constructs.Construct, id: string) {
    super(scope, id);

    this.provisionedProduct = new sc.CfnCloudFormationProvisionedProduct(this, 'Resource', {
      provisionedProductName: this.node.id,
      provisioningArtifactName: 'v1.0',
      productName: 'AI-Powered Health Data Masking',
    });

    this.provisionedProductId = this.provisionedProduct.ref;
    this.cloudFormationStackArn = cdk.Token.asString(cdk.Fn.getAtt(this.provisionedProduct.logicalId, 'CloudformationStackArn'));
    this.recordId = cdk.Token.asString(cdk.Fn.getAtt(this.provisionedProduct.logicalId, 'RecordId'));
    this.apiGatewayInvokeURL = cdk.Token.asString(cdk.Fn.getAtt(this.provisionedProduct.logicalId, 'Outputs.ApiGatewayInvokeURL'));
    this.apiGatewayId = cdk.Token.asString(cdk.Fn.getAtt(this.provisionedProduct.logicalId, 'Outputs.ApiGatewayId'));
    this.apiGatewayStage = cdk.Token.asString(cdk.Fn.getAtt(this.provisionedProduct.logicalId, 'Outputs.ApiGatewayStage'));
    this.imageBucketName = cdk.Token.asString(cdk.Fn.getAtt(this.provisionedProduct.logicalId, 'Outputs.ImageBucketName'));
    this.logBucketName = cdk.Token.asString(cdk.Fn.getAtt(this.provisionedProduct.logicalId, 'Outputs.LogBucketName'));
    this.imageResourceId = cdk.Token.asString(cdk.Fn.getAtt(this.provisionedProduct.logicalId, 'Outputs.ImageResourceId'));
    this.imageTextResourceId = cdk.Token.asString(cdk.Fn.getAtt(this.provisionedProduct.logicalId, 'Outputs.ImageTextResourceId'));
    this.imagePhiResourceId = cdk.Token.asString(cdk.Fn.getAtt(this.provisionedProduct.logicalId, 'Outputs.ImagePhiResourceId'));
    this.imageMaskResourceId = cdk.Token.asString(cdk.Fn.getAtt(this.provisionedProduct.logicalId, 'Outputs.ImageMaskResourceId'));
    this.textPhiResourceId = cdk.Token.asString(cdk.Fn.getAtt(this.provisionedProduct.logicalId, 'Outputs.TextPhiResourceId'));
    this.textMaskResourceId = cdk.Token.asString(cdk.Fn.getAtt(this.provisionedProduct.logicalId, 'Outputs.TextMaskResourceId'));
  }
}"
`;

exports[`generate bucket.json 1`] = `
"// Generated by cdk-import
import * as cdk from 'aws-cdk-lib';
import * as constructs from 'constructs';
import * as sc from 'aws-cdk-lib/aws-servicecatalog';

export interface IBucketProduct extends cdk.IResource {
  readonly provisionedProductId: string;
  readonly cloudFormationStackArn: string;
  readonly recordId: string;
}

abstract class BucketProductBase extends cdk.Resource implements IBucketProduct {
  public abstract readonly provisionedProductId: string;
  public abstract readonly cloudFormationStackArn: string;
  public abstract readonly recordId: string;
}

/**
 * A Service Catalog CloudFormation \`BucketProduct\` product.
 *
 * @cloudformationResource AWS::ServiceCatalog::CloudFormationProvisionedProduct
 * @stability external
 */
export class BucketProduct extends BucketProductBase {
  /**
   * Attribute \`BucketProduct.provisionedProductId\`
   */
  public readonly provisionedProductId: string;
  /**
   * Attribute \`BucketProduct.cloudFormationStackArn\`
   */
  public readonly cloudFormationStackArn: string;
  /**
   * Attribute \`BucketProduct.recordId\`
   */
  public readonly recordId: string;
  private readonly provisionedProduct: sc.CfnCloudFormationProvisionedProduct;

  /**
   * Create a new Service Catalog \`BucketProduct\` product.
   *
   * @param scope - scope in which this resource is defined
   * @param id    - scoped id of the resource
   */
  constructor(scope: constructs.Construct, id: string) {
    super(scope, id);

    this.provisionedProduct = new sc.CfnCloudFormationProvisionedProduct(this, 'Resource', {
      provisionedProductName: this.node.id,
      provisioningArtifactName: 'BucketProduct',
      productName: 'BucketProduct',
    });

    this.provisionedProductId = this.provisionedProduct.ref;
    this.cloudFormationStackArn = cdk.Token.asString(cdk.Fn.getAtt(this.provisionedProduct.logicalId, 'CloudformationStackArn'));
    this.recordId = cdk.Token.asString(cdk.Fn.getAtt(this.provisionedProduct.logicalId, 'RecordId'));
  }
}"
`;

exports[`generate ec2.json 1`] = `
"// Generated by cdk-import
import * as cdk from 'aws-cdk-lib';
import * as constructs from 'constructs';
import * as sc from 'aws-cdk-lib/aws-servicecatalog';

export interface IEc2ComputeInstance extends cdk.IResource {
  readonly provisionedProductId: string;
  readonly cloudFormationStackArn: string;
  readonly recordId: string;
  readonly iPAddress: string;
  readonly downloadKeyCommand: string;
  readonly sshcommand: string;
}

abstract class Ec2ComputeInstanceBase extends cdk.Resource implements IEc2ComputeInstance {
  public abstract readonly provisionedProductId: string;
  public abstract readonly cloudFormationStackArn: string;
  public abstract readonly recordId: string;
  public abstract readonly iPAddress: string;
  public abstract readonly downloadKeyCommand: string;
  public abstract readonly sshcommand: string;
}

enum ParameterKeys {
  ssmParameterValueawsserviceamiamazonlinuxlatestamzn2amihvmarm64gp2C96584B6F00A464EAD1953AFF4B05118Parameter = 'SsmParameterValueawsserviceamiamazonlinuxlatestamzn2amihvmarm64gp2C96584B6F00A464EAD1953AFF4B05118Parameter',
  instanceType = 'InstanceType',
}

namespace ParameterKeys {
  const reverseMap = new Map<string, ParameterKeys>();
  Object.keys(ParameterKeys).forEach((s: string) => {
    const e = (<any>ParameterKeys)[s];
    reverseMap.set(s, e);
  });
  export function valueOf(str: string) {
    return reverseMap.get(str);
  }
}
export interface Ec2ComputeInstanceProps {
  /**
   *
   * @default /aws/service/ami-amazon-linux-latest/amzn2-ami-hvm-arm64-gp2
   */
  readonly ssmParameterValueawsserviceamiamazonlinuxlatestamzn2amihvmarm64gp2C96584B6F00A464EAD1953AFF4B05118Parameter?: string;

  /**
   * The instance type of an EC2 instance.
   */
  readonly instanceType: string;

}

/**
 * A Service Catalog CloudFormation \`Ec2ComputeInstance\` product.
 *
 * @cloudformationResource AWS::ServiceCatalog::CloudFormationProvisionedProduct
 * @stability external
 */
export class Ec2ComputeInstance extends Ec2ComputeInstanceBase {
  /**
   * Attribute \`Ec2ComputeInstance.provisionedProductId\`
   */
  public readonly provisionedProductId: string;
  /**
   * Attribute \`Ec2ComputeInstance.cloudFormationStackArn\`
   */
  public readonly cloudFormationStackArn: string;
  /**
   * Attribute \`Ec2ComputeInstance.recordId\`
   */
  public readonly recordId: string;
  /**
   * Attribute \`Ec2ComputeInstance.iPAddress\`
   */
  public readonly iPAddress: string;
  /**
   * Attribute \`Ec2ComputeInstance.downloadKeyCommand\`
   */
  public readonly downloadKeyCommand: string;
  /**
   * Attribute \`Ec2ComputeInstance.sshcommand\`
   */
  public readonly sshcommand: string;
  private readonly provisionedProduct: sc.CfnCloudFormationProvisionedProduct;

  /**
   * Create a new Service Catalog \`Ec2ComputeInstance\` product.
   *
   * @param scope - scope in which this resource is defined
   * @param id    - scoped id of the resource
   * @param props - resource properties
   */
  constructor(scope: constructs.Construct, id: string, props: Ec2ComputeInstanceProps) {
    super(scope, id);

    this.provisionedProduct = new sc.CfnCloudFormationProvisionedProduct(this, 'Resource', {
      provisionedProductName: this.node.id,
      provisioningArtifactName: 'EC2LargeInstance',
      productName: 'Ec2ComputeInstance',
      provisioningParameters:
        Object.entries(props).map(([k, v]) => {
          return {
            key: ParameterKeys.valueOf(k) as string,
            value: v as string,
          }
        }),
    });

    this.provisionedProductId = this.provisionedProduct.ref;
    this.cloudFormationStackArn = cdk.Token.asString(cdk.Fn.getAtt(this.provisionedProduct.logicalId, 'CloudformationStackArn'));
    this.recordId = cdk.Token.asString(cdk.Fn.getAtt(this.provisionedProduct.logicalId, 'RecordId'));
    this.iPAddress = cdk.Token.asString(cdk.Fn.getAtt(this.provisionedProduct.logicalId, 'Outputs.IPAddress'));
    this.downloadKeyCommand = cdk.Token.asString(cdk.Fn.getAtt(this.provisionedProduct.logicalId, 'Outputs.DownloadKeyCommand'));
    this.sshcommand = cdk.Token.asString(cdk.Fn.getAtt(this.provisionedProduct.logicalId, 'Outputs.sshcommand'));
  }
}"
`;
